File: CS50x/Week06 - Python/Sentimental-Credit/credit.py
from cs50 import get_string


def main():
    # Prompt user for credit card number with validation
    while True:
        number = get_string("Number: ")
        if number.isnumeric() and len(number) <= 16:
            break

    # Conditions to verify if the cards are valid and, if so, from witch company

    # VISA -> Has to have 13 or 16 digits, the first digit must be 4
    if luhns(number) and (len(number) == 13 or len(number) == 16) and number[0] == '4':
        print("VISA")

    # American Express -> Has to have 15 digits, first two digits must be 34 or 37
    elif luhns(number) and len(number) == 15 and number[0] == '3' and (number[1] == '4' or number[1] == '7'):
        print("AMEX")

    # Master Card -> Has to have 16 digits, the first two digits must be 51, 52, 53, 54 or 55
    elif luhns(number) and len(number) == 16 and number[0] == '5' and (number[1] in ['1', '2', '3', '4', '5']):
        print("MASTERCARD")

    else:
        print("INVALID")


def luhns(num):
    doubles_sum = 0
    # Second-to-last digit and so on
    for i in range(len(num) - 2, -1, -2):
        # Multiplied by 2
        digit_double = int(num[i]) * 2

        # Checking if number > 9 to sum the digits
        if digit_double > 9:
            # Converts number to string
            s = str(digit_double)

            # Adds those numbers
            digit_double = int(s[0]) + int(s[1])

        doubles_sum += digit_double

    # Leftover digits sum + doubled sum
    lefts_sum = 0
    for i in range(len(num) - 1, -1, -2):
        lefts_sum += int(num[i])

    # Final Sum
    sum = doubles_sum + lefts_sum

    if (sum % 10 == 0):
        return True
    else:
        return False


main()

File: CS50x/Week01 - C/Credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    int countMain = 0;
    int countAux = 0;
    int sumOthers = 0;
    int sumLeftovers = 0;

    // Prompt user for credit card number
    long number = get_long("Number: ");

    // Getting the "other digits" of the number
    for (long i = 1; i < number; i *= 10)
    {
        // Only counting the requested "other digits"
        if (countMain % 2 != 0)
        {
            // Multiplying those numbers by 2
            int digit = ((number / i) % 10) * 2;
            // Breaking the digits bigger than 9 to sum them with each other
            if (digit > 9)
            {
                digit = (digit / 10) % 10 + digit % 10;
            }
            // Summing those digits
            sumOthers += digit;
        }
        // Counts the digits of the given number
        countMain++;
    }

    // Geting first digit
    long quoefFirstDigit = pow(10, countMain - 1);
    int firstDigit = (number / quoefFirstDigit) % 10;
    // Getting second digit
    long quoefSecondDigit = pow(10, countMain - 2);
    int secondDigit = (number / quoefSecondDigit) % 10;

    // Geting the leftover digits
    for (long i = 1; i < number; i *= 10)
    {
        // Here we have the even indentations, contrarily to the other ones, witch was counting the
        // odd iterations
        if (countAux % 2 == 0)
        {
            int digit = (number / i) % 10;
            // Summing the digits
            sumLeftovers += digit;
        }
        countAux++;
    }
    // For the number to be valid, this total sum has to be a 10 potency
    int module = (sumLeftovers + sumOthers) % 10;

    // Conditions to verify if the cards are valid and, if so, from witch company

    // VISA -> Has to have 13 or 16 digits, the first digit must be 4, module has to be 0
    if ((countMain == 13 || countMain == 16) && firstDigit == 4 && module == 0)
    {
        printf("VISA\n");
    }

    // American Express -> Has to have 15 digits, first two digits must be 34 or 37, module has to
    // be 0
    else if (countMain == 15 && firstDigit == 3 && (secondDigit == 4 || secondDigit == 7) &&
             module == 0)
    {
        printf("AMEX\n");
    }

    // Master Card -> Has to have 16 digits, the first two digits must be 51, 52, 53, 54 or 55,
    // module has to be 0
    else if (countMain == 16 && firstDigit == 5 &&
             (secondDigit == 1 || secondDigit == 2 || secondDigit == 3 || secondDigit == 4 ||
              secondDigit == 5) &&
             module == 0)
    {
        printf("MASTERCARD\n");
    }

    else
    {
        printf("INVALID\n");
    }
}
